### PR TITLE
cli/install.php - Add command-line interface to the installation system

### PR DESCRIPTION
**Pull request by totten**
Example usage:

``` bash
git clone https://github.com/joomla/joomla-cms.git
cd joomla-cms
php cli/install.php \
   --db-user=myuser --db-name=mydb --db-pass=mysecret \
   --admin-user=admin --admin-pass=secret --admin-email='admin@example.com'
rm -rf installation
```

This is my first patch to Joomla, and the patch has a couple FIXME's which may require discussion. The main FIXMEs:
- JApplicationCliInstaller::doExecute() sets "$options['db_created'] = 1" as a workaround. This shouldn't be necessary because InstallationModelDatabase tries to set that value by way of $session, but something's wonky with the handling of $session, and I don't know the ins-and-outs about JApplicationCli/JApplicationCms/JSession.
- getLocaliseAdmin() has been brutishly copied-and-pasted from InstallationApplicationWeb to JApplicationCliInstaller. I don't particularly know what it's for, but it doesn't appear to depend on $this, so it could be safely moved to another class (where it's easier to share code).

### CODE REVIEW

File: cli/install.php [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+	 * Returns the installed language files in the administrative and
+	 * front-end area.
+	 *
+	 * @param   mixed  $db  JDatabaseDriver instance.
+	 *
+	 * @return  array  Array with installed language packs in admin and site area.
+	 *
+	 * @since   3.4
+	 */
+	public function getLocaliseAdmin($db = false)
```
Comment (totten) [inline cli/install.php]: This is one of my initial FIXMEs - it doesn't make sense to me.

The getLocaliseAdmin() is originally defined in installation/application/web.php and had to be replicated in cli/install.php . It is called by installation/model/database.php . It seems weird to me that a model-class would depend on the application-class, but I'm new here... so [I don't know what I don't know](http://www.amazon.com/Pieces-Intelligence-Existential-Poetry-Rumsfeld/dp/1439167230) and it's hard for me to say whether the model needs refactoring. As a new contrib, I tried to keep change focused in the new parts.

**Comment by colemanw** [comment]
This proposed improvement is a really exciting step forward for Joomla. It sounds like there a few minor things about it that could be improved (most of which totten has already addressed) but overall it works and is very useful.
Could someone please either merge this PR or explain why not?

**Comment by nikosdion** [comment]
@colemanw I have also written a similar CLi script and filed a PR. While discussing it we figured out that JInstaller and many 3PD extensions' installation scripts depend on methods which are only present in the JApplicationWeb / JApplicationAdministrator but not in JApplicationCli which is the only application class which will work properly under CLI. Until JInstaller is fully decoupled from JApplicationAdministrator we can't have a CLI extensions installer script because it won't work with many (most?) third party extensions.

**Comment by nikosdion** [comment]
@colemanw It's not obvious from what I said: installing Joomla! also bumps on the similar obstacles as my extensions installation PR, that's why *BOTH* PRs are currently not merged.

**Comment by colemanw** [comment]
@nikosdion thanks for the explanation.

**Comment by totten** [comment]
On one hand, I agree with @nikosdion that this stumbled slightly on the issue of JApplication(Web vs Cli) -- which basically led to the small kludges/workarounds highlighted in the original description. OTOH, I think it's less problematic for the core-installer than for the extension-installer. (The core installer is a smaller, more closed use-case with everything defined in the core repo.) From my perspective, the main obstacle was more managerial -- not knowing which of the two kludges needed to be addressed to make it mergable (or even if the feature generally was considered acceptable).

Regardless, @stevenrombauts appears to be making progress on this use-case as part of joomla-console (https://github.com/joomlatools/joomla-console/blob/develop/src/Joomlatools/Console/Command/Site/Install.php , https://github.com/joomlatools/joomla-console/pull/5), so I'm closing the PR on joomla-cms.

**Comment by btoplak** [comment]
good thing I have found this ticket. I've already started working on my own similar code, but it's best to join all of our efforts on the same project

**Pull request closed without merging**